Compilation Exception due to ASM?

Running on clojure-1.5.1
cc @jeffvalk 
Error loading cider.nrepl.middleware.info: java.lang.RuntimeException: Unable to find static field: ASM4 in interface org.objectweb.asm.Opcodes, compiling:(cider/nrepl/middleware/util/java.clj:68:17)
Exception in thread "main" java.lang.RuntimeException: Unable to resolve var: cider.nrepl.middleware.info/wrap-info in this context, compiling:(/tmp/form-init7656399187810875460.clj:1:3946)
    at clojure.lang.Compiler.analyzeSeq(Compiler.java:6567)
    at clojure.lang.Compiler.analyze(Compiler.java:6361)
    at clojure.lang.Compiler.analyze(Compiler.java:6322)
    at clojure.lang.Compiler$InvokeExpr.parse(Compiler.java:3624)
    at clojure.lang.Compiler.analyzeSeq(Compiler.java:6562)
    at clojure.lang.Compiler.analyze(Compiler.java:6361)
    at clojure.lang.Compiler.analyze(Compiler.java:6322)
    at clojure.lang.Compiler$InvokeExpr.parse(Compiler.java:3624)
    at clojure.lang.Compiler.analyzeSeq(Compiler.java:6562)
    at clojure.lang.Compiler.analyze(Compiler.java:6361)
    at clojure.lang.Compiler.access$100(Compiler.java:37)
    at clojure.lang.Compiler$LetExpr$Parser.parse(Compiler.java:5973)
    at clojure.lang.Compiler.analyzeSeq(Compiler.java:6560)
    at clojure.lang.Compiler.analyze(Compiler.java:6361)
    at clojure.lang.Compiler.analyze(Compiler.java:6322)
    at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:5708)
    at clojure.lang.Compiler$FnMethod.parse(Compiler.java:5139)
    at clojure.lang.Compiler$FnExpr.parse(Compiler.java:3751)
    at clojure.lang.Compiler.analyzeSeq(Compiler.java:6558)
    at clojure.lang.Compiler.analyze(Compiler.java:6361)
    at clojure.lang.Compiler.eval(Compiler.java:6616)
    at clojure.lang.Compiler.eval(Compiler.java:6609)
    at clojure.lang.Compiler.load(Compiler.java:7064)
    at clojure.lang.Compiler.loadFile(Compiler.java:7020)
    at clojure.main$load_script.invoke(main.clj:294)
    at clojure.main$init_opt.invoke(main.clj:299)
    at clojure.main$initialize.invoke(main.clj:327)
    at clojure.main$null_opt.invoke(main.clj:362)
    at clojure.main$main.doInvoke(main.clj:440)
    at clojure.lang.RestFn.invoke(RestFn.java:421)
    at clojure.lang.Var.invoke(Var.java:419)
    at clojure.lang.AFn.applyToHelper(AFn.java:163)
    at clojure.lang.Var.applyTo(Var.java:532)
    at clojure.main.main(main.java:37)
Caused by: java.lang.RuntimeException: Unable to resolve var: cider.nrepl.middleware.info/wrap-info in this context
    at clojure.lang.Util.runtimeException(Util.java:219)
    at clojure.lang.Compiler$TheVarExpr$Parser.parse(Compiler.java:650)
    at clojure.lang.Compiler.analyzeSeq(Compiler.java:6560)
    ... 33 more
Subprocess failed

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,5 +1,9 @@
 language: clojure
-script: lein2 with-profile dev test
+script:
+  - lein2 with-profile dev test
+  - lein2 with-profile dev,1.5 test
+  - lein2 with-profile dev,1.6 test
+  - lein2 with-profile dev,1.7 test
 jdk:
   - openjdk6
   - openjdk7

--- a/project.clj
+++ b/project.clj
@@ -6,14 +6,15 @@
   :dependencies [[org.clojure/clojure "1.5.1"]
                  [compliment "0.0.3"]
                  [cljs-tooling "0.1.2"]
-                 [org.ow2.asm/asm "5.0.2"]
-                 [org.ow2.asm/asm-commons "5.0.2"]
                  [org.tcrawley/dynapath "0.2.3"]
                  [org.clojure/tools.nrepl "0.2.3"]
                  [org.clojure/java.classpath "0.2.0"]
                  [org.clojure/tools.namespace "0.2.3"]]
 
-  :profiles {:dev {:repl-options {:nrepl-middleware [cider.nrepl.middleware.classpath/wrap-classpath
+  :profiles {:1.5 {:dependencies [[org.clojure/clojure "1.5.1"]]}
+             :1.6 {:dependencies [[org.clojure/clojure "1.6.0"]]}
+             :1.7 {:dependencies [[org.clojure/clojure "1.7.0-master-SNAPSHOT"]]}
+             :dev {:repl-options {:nrepl-middleware [cider.nrepl.middleware.classpath/wrap-classpath
                                                      cider.nrepl.middleware.complete/wrap-complete
                                                      cider.nrepl.middleware.info/wrap-info
                                                      cider.nrepl.middleware.inspect/wrap-inspect

--- a/src/cider/nrepl/middleware/util/java.clj
+++ b/src/cider/nrepl/middleware/util/java.clj
@@ -4,8 +4,8 @@
   (:require [clojure.java.io :as io]
             [clojure.string :as str]
             [dynapath.util :as dp])
-  (:import (org.objectweb.asm ClassReader ClassVisitor MethodVisitor Opcodes)
-           (org.objectweb.asm.commons Method)))
+  (:import (clojure.asm ClassReader ClassVisitor MethodVisitor Opcodes)
+           (clojure.asm.commons Method)))
 
 ;;; ## Source Files
 ;; Java source files are resolved from the classpath. For library dependencies,
@@ -58,6 +58,26 @@
 ;; executable line of the method, rather than its declaration. (Constructors are
 ;; an exception to this.)
 
+;; ASM Workaround: Clojure <=1.5 bundles ASM 3.3; Clojure 1.6 upgraded to ASM
+;; 4.1, which had breaking API changes. Declaring our own ASM dependency causes
+;; headaches (see issue #44), so we'll work around the API changes.
+(def asm
+  "The ASM API version"
+  (try (import (clojure.asm.commons EmptyVisitor)) 3
+       (catch Exception _ 4)))
+
+;; In the ASM4 API, visitors are abstract classes with stub methods whose
+;; constructrors require the API version as an argument. In ASM3, visitor were
+;; interfaces, and stubs were provided by the `EmptyVisitor` class, which has
+;; only a nullary constructor.
+(defmacro proxy-asm
+  "Like `proxy`, but provides a shim to support both ASM3 and ASM4 visitors."
+  [supers args & fs]
+  (let [[supers args] (if (< asm 4)
+                        [`[EmptyVisitor] []]
+                        [supers `[Opcodes/ASM4]])]
+    `(proxy ~supers ~args ~@fs)))
+
 (defn class-info
   "For the named class, return Java source and member info, including line
   numbers. Methods are indexed first by name, and then by argument types to
@@ -66,7 +86,7 @@
   (let [methods (atom {})
         parent  (atom nil)
         typesym #(-> % .getClassName symbol)
-        visitor (proxy [ClassVisitor] [Opcodes/ASM4]
+        visitor (proxy-asm [ClassVisitor] []
                   (visit [version access name signature super interfaces]
                     (when super
                       (let [super (str/replace super "/" ".")]
@@ -75,7 +95,7 @@
                     (let [m (Method. name desc)
                           ret (typesym (.getReturnType m))
                           args (mapv typesym (.getArgumentTypes m))]
-                      (proxy [MethodVisitor] [Opcodes/ASM4]
+                      (proxy-asm [MethodVisitor] []
                         (visitLineNumber [line label]
                           (when-not (get-in @methods [name args])
                             (swap! methods assoc-in [name args]

